src/typescript-service.ts

```diff
@@ -338,6 +338,847 @@ export default class TypeScriptService {
 		});
 	}
 
+	getDocumentSymbol(params: DocumentSymbolParams): Promise<SymbolInformation[]> {
+		const uri = params.textDocument.uri;
+		return this.ensureFilesForHoverAndDefinition(uri).then(() => {
+			const fileName = util.uri2path(uri);
+			const config = this.projectManager.getConfiguration(uri);
+			return config.prepare(this.connection).then(() => {
+				this.projectManager.syncConfiguration(config, this.connection);
+				const sourceFile = config.program.getSourceFile(fileName);
+				const tree = config.service.getNavigationTree(fileName);
+				const limit = 100000;
+				const result: SymbolInformation[] = [];
+				this.flattenNavigationTreeItem(tree, null, sourceFile, result, limit);
+				return Promise.resolve(result);
+			});
+		});
+	}
+
+	getWorkspaceReference(params: util.WorkspaceReferenceParams): Promise<util.ReferenceInformation[]> {
+		const refInfo: util.ReferenceInformation[] = [];
+		return this.ensureFilesForWorkspaceSymbol().then(() => {
+			return Promise.all(this.projectManager.getConfigurations().map((config) => {
+				return config.prepare(this.connection).then((config) => {
+					this.projectManager.syncConfiguration(config, this.connection);
+					for (let source of config.service.getProgram().getSourceFiles()) {
+						if (source.fileName.indexOf("/node_modules/") !== -1) {
+							continue;
+						}
+						this.walkMostAST(source, (node) => {
+							switch (node.kind) {
+								case ts.SyntaxKind.Identifier: {
+									const fileName = source.fileName.startsWith("/") ? source.fileName.substr(1) : source.fileName;
+									const id = node as ts.Identifier;
+									const defs = config.service.getDefinitionAtPosition(fileName, node.pos + 1);
+
+									if (defs && defs.length > 0) {
+										const def = defs[0];
+										const start = ts.getLineAndCharacterOfPosition(source, node.pos);
+										const end = ts.getLineAndCharacterOfPosition(source, node.end);
+										const ref = {
+											location: {
+												uri: util.path2uri('', source.fileName),
+												range: {
+													start: start,
+													end: end,
+												},
+											},
+											name: def.name,
+											containerName: def.containerName,
+											uri: util.path2uri('', def.fileName),
```

Comment:
use `this.defUri` instead to deal with TS-originated defs. Also, please find an extract from results on Windows in non-strict mode (from vscode):
```
  { location:
     { uri: 'file:///c:/projects/langserver-typescript/src/util.ts',
       range: [Object] },
    name: 'ts',
    containerName: '',
    uri: 'file:///c:/projects/langserver-typescript/src/util.ts' },
```
Note uris here